File: src/recall/integrity.ts
/**
 * Hash chain — canonical JSON, sha256, link verification.
 *
 * The chain is the only thing standing between an honest operator and
 * an operator who quietly rewrites yesterday's losing spin into a win.
 * Two invariants:
 *
 *   1. `entry_hash` = sha256(canonical_json(entry without `entry_hash`)).
 *   2. `prev_hash`  = previous row's `entry_hash`.
 *
 * Break either invariant and verification fails noisily.
 *
 * The Rust mirror in `rust-sim/src/recall/integrity.rs` produces the
 * same bytes for the same input — `tests/fixtures/recall/kat-chain.json`
 * is the cross-language KAT.
 */

import { createHash } from 'crypto';
import type { Hex64, JournalManifest, SpinJournalEntry } from './types.js';
import { RECALL_SCHEMA_VERSION, ZERO_HASH } from './types.js';

/**
 * Canonical JSON for hashing. Two rules:
 *   - Keys sorted lexicographically at every level.
 *   - No whitespace.
 * The ECMA spec leaves object-property order to the implementation, so
 * we sort explicitly. Numbers go through `JSON.stringify(num)` which
 * already produces the shortest stable form for finite numbers and
 * refuses NaN / Infinity (returns `null`, which we treat as a write-
 * time error elsewhere; here we just trust it because the journal
 * writer guards finiteness).
 */
export function canonicalJson(value: unknown): string {
  if (value === null) return 'null';
  if (typeof value === 'string') return JSON.stringify(value);
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) {
      throw new Error(`canonicalJson: refusing to serialize non-finite ${value}`);
    }
    return JSON.stringify(value);
  }
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  if (Array.isArray(value)) {
    return '[' + value.map(canonicalJson).join(',') + ']';
  }
  if (typeof value === 'object') {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    return (
      '{' +
      keys.map((k) => JSON.stringify(k) + ':' + canonicalJson(obj[k])).join(',') +
      '}'
    );
  }
  throw new Error(`canonicalJson: unsupported type ${typeof value}`);
}

/** sha256 of a UTF-8 string, lowercase hex. */
export function sha256Hex(s: string): Hex64 {
  return createHash('sha256').update(s, 'utf8').digest('hex');
}

/**
 * Compute the `entry_hash` of a journal entry, given the entry WITHOUT
 * its `entry_hash` field. Caller fills the field with this return value
 * before writing.
 */
export function computeEntryHash(entryWithoutHash: Omit<SpinJournalEntry, 'entry_hash'>): Hex64 {
  return sha256Hex(canonicalJson(entryWithoutHash as unknown));
}

/**
 * Compute the `manifest_hash` of a manifest, given the manifest WITHOUT
 * its `manifest_hash` field.
 */
export function computeManifestHash(
  manifestWithoutHash: Omit<JournalManifest, 'manifest_hash'>,
): Hex64 {
  return sha256Hex(canonicalJson(manifestWithoutHash as unknown));
}

/**
 * Stamp the entry: fill in `prev_hash` from the head and compute
 * `entry_hash`. Mutates a *copy*, returns the finalized entry.
 *
 * The head argument is the running tail hash of the chain — the writer
 * keeps it as state. `null` ⇒ first entry, uses `ZERO_HASH`.
 */
export function sealEntry(
  draft: Omit<SpinJournalEntry, 'prev_hash' | 'entry_hash'>,
  head: Hex64 | null,
): SpinJournalEntry {
  const prev_hash = head ?? ZERO_HASH;
  const withPrev: Omit<SpinJournalEntry, 'entry_hash'> = { ...draft, prev_hash };
  const entry_hash = computeEntryHash(withPrev);
  return { ...withPrev, entry_hash };
}

/**
 * Stamp the manifest: compute `manifest_hash` against the manifest
 * minus that field.
 */
export function sealManifest(
  draft: Omit<JournalManifest, 'manifest_hash'>,
): JournalManifest {
  const manifest_hash = computeManifestHash(draft);
  return { ...draft, manifest_hash };
}

export type ChainVerification =
  | { ok: true; count: number; last_entry_hash: Hex64 }
  | {
      ok: false;
      reason:
        | 'empty_chain'
        | 'genesis_prev_not_zero'
        | 'prev_hash_mismatch'
        | 'entry_hash_mismatch'
        | 'seq_not_monotonic'
        | 'schema_version_mismatch';
      seq: number | null;
      detail: string;
    };

/**
 * Verify a chain in memory: recompute each entry's hash, check `prev_hash`
 * links, check `seq` is monotonic. Returns the first failure noisily.
 *
 * Empty chain is an error — callers verifying a journal they expected
 * to contain entries should not silently accept an empty list.
 */
export function verifyChain(entries: SpinJournalEntry[]): ChainVerification {
  if (entries.length === 0) {
    return {
      ok: false,
      reason: 'empty_chain',
      seq: null,
      detail: 'verifyChain called with zero entries',
    };
  }
  let prevHash: Hex64 = ZERO_HASH;
  let prevSeq = -1;
  for (const e of entries) {
    if (e.schema_version !== RECALL_SCHEMA_VERSION) {
      return {
        ok: false,
        reason: 'schema_version_mismatch',
        seq: e.seq,
        detail: `expected ${RECALL_SCHEMA_VERSION}, got ${e.schema_version}`,
      };
    }
    if (e.seq <= prevSeq) {
      return {
        ok: false,
        reason: 'seq_not_monotonic',
        seq: e.seq,
        detail: `seq ${e.seq} ≤ previous ${prevSeq}`,
      };
    }
    if (e.prev_hash !== prevHash) {
      return {
        ok: false,
        reason: prevSeq === -1 ? 'genesis_prev_not_zero' : 'prev_hash_mismatch',
        seq: e.seq,
        detail: `expected prev_hash ${prevHash}, got ${e.prev_hash}`,
      };
    }
    // Recompute the hash.
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { entry_hash: _ignored, ...withoutHash } = e;
    const want = computeEntryHash(withoutHash);
    if (want !== e.entry_hash) {
      return {
        ok: false,
        reason: 'entry_hash_mismatch',
        seq: e.seq,
        detail: `recomputed ${want}, stored ${e.entry_hash}`,
      };
    }
    prevHash = e.entry_hash;
    prevSeq = e.seq;
  }
  return { ok: true, count: entries.length, last_entry_hash: prevHash };
}

File: rust-sim/src/recall/integrity.rs
//! Canonical-JSON + sha256 chain — Rust mirror of `src/recall/integrity.ts`.
//!
//! The canonical-JSON serializer in particular MUST byte-match the TS
//! side, or the cross-language KAT will fail (intentional design — that's
//! exactly what catches drift before it ends up in a regulator's audit
//! mailbox).

use super::types::*;
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::io::Write;

/// Serialize a `serde_json::Value` to canonical JSON: keys sorted at
/// every nesting level, no whitespace, no trailing newline. Numbers go
/// through `serde_json`'s display so finite f64 share the shortest stable
/// representation with TS's `JSON.stringify`. Non-finite numbers are
/// already forbidden by serde_json on the deserialize path.
pub fn canonical_json(value: &Value) -> String {
    let mut buf = Vec::with_capacity(256);
    write_canonical(&mut buf, value);
    // SAFETY: `write_canonical` only emits valid UTF-8.
    String::from_utf8(buf).expect("canonical_json produced invalid UTF-8")
}

fn write_canonical(out: &mut Vec<u8>, value: &Value) {
    match value {
        Value::Null => out.extend_from_slice(b"null"),
        Value::Bool(b) => out.extend_from_slice(if *b { b"true" } else { b"false" }),
        Value::Number(n) => {
            // Reject non-finite — `serde_json::Number` cannot hold NaN/Inf
            // for f64, so this branch is effectively defensive. We still
            // assert to surface a logic bug noisily if it ever changes.
            if let Some(f) = n.as_f64() {
                assert!(
                    f.is_finite(),
                    "canonical_json: non-finite f64 leaked into Value"
                );
            }
            write!(out, "{}", n).unwrap();
        }
        Value::String(s) => {
            // Reuse serde_json's string serializer — it produces the same
            // escape sequence set as JS `JSON.stringify`.
            let s = serde_json::to_string(s).unwrap();
            out.extend_from_slice(s.as_bytes());
        }
        Value::Array(arr) => {
            out.push(b'[');
            for (i, item) in arr.iter().enumerate() {
                if i > 0 {
                    out.push(b',');
                }
                write_canonical(out, item);
            }
            out.push(b']');
        }
        Value::Object(map) => {
            // Sort keys lexicographically.
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            out.push(b'{');
            for (i, k) in keys.iter().enumerate() {
                if i > 0 {
                    out.push(b',');
                }
                let key_json = serde_json::to_string(k).unwrap();
                out.extend_from_slice(key_json.as_bytes());
                out.push(b':');
                write_canonical(out, &map[*k]);
            }
            out.push(b'}');
        }
    }
}

/// sha256 of a UTF-8 string, lowercase hex.
pub fn sha256_hex(s: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(s.as_bytes());
    let digest = hasher.finalize();
    let mut out = String::with_capacity(64);
    for b in digest.iter() {
        // hex-format byte with leading zero
        out.push_str(&format!("{:02x}", b));
    }
    out
}

/// Compute `entry_hash` for an entry that doesn't yet have one. The TS
/// counterpart strips the `entry_hash` field from the object before
/// canonicalizing; here we emit the entry via serde, remove the field
/// from the resulting `Value`, then canonicalize.
pub fn compute_entry_hash(entry: &SpinJournalEntry) -> String {
    let mut v = serde_json::to_value(entry).expect("entry serializes");
    if let Some(obj) = v.as_object_mut() {
        obj.remove("entry_hash");
    }
    sha256_hex(&canonical_json(&v))
}

/// Compute `manifest_hash` over the manifest minus that field.
pub fn compute_manifest_hash(manifest: &JournalManifest) -> String {
    let mut v = serde_json::to_value(manifest).expect("manifest serializes");
    if let Some(obj) = v.as_object_mut() {
        obj.remove("manifest_hash");
    }
    sha256_hex(&canonical_json(&v))
}

/// Builder helper: fill `prev_hash` from the running chain head and
/// compute `entry_hash`. Returns the finalized entry.
pub fn seal_entry(mut draft: SpinJournalEntry, head: Option<&str>) -> SpinJournalEntry {
    draft.prev_hash = head.unwrap_or(ZERO_HASH).to_string();
    draft.entry_hash = String::new(); // ignored by compute_entry_hash
    let hash = compute_entry_hash(&draft);
    draft.entry_hash = hash;
    draft
}

/// Builder helper for the manifest.
pub fn seal_manifest(mut draft: JournalManifest) -> JournalManifest {
    draft.manifest_hash = String::new();
    let hash = compute_manifest_hash(&draft);
    draft.manifest_hash = hash;
    draft
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ChainVerification {
    Ok {
        count: usize,
        last_entry_hash: String,
    },
    Empty,
    GenesisPrevNotZero {
        seq: u64,
        detail: String,
    },
    PrevHashMismatch {
        seq: u64,
        detail: String,
    },
    EntryHashMismatch {
        seq: u64,
        detail: String,
    },
    SeqNotMonotonic {
        seq: u64,
        detail: String,
    },
    SchemaVersionMismatch {
        seq: u64,
        detail: String,
    },
}

pub fn verify_chain(entries: &[SpinJournalEntry]) -> ChainVerification {
    if entries.is_empty() {
        return ChainVerification::Empty;
    }
    let mut prev_hash = ZERO_HASH.to_string();
    let mut prev_seq: i64 = -1;
    for e in entries {
        if e.schema_version != RECALL_SCHEMA_VERSION {
            return ChainVerification::SchemaVersionMismatch {
                seq: e.seq,
                detail: format!(
                    "expected {}, got {}",
                    RECALL_SCHEMA_VERSION, e.schema_version
                ),
            };
        }
        if (e.seq as i64) <= prev_seq {
            return ChainVerification::SeqNotMonotonic {
                seq: e.seq,
                detail: format!("seq {} ≤ previous {}", e.seq, prev_seq),
            };
        }
        if e.prev_hash != prev_hash {
            return if prev_seq == -1 {
                ChainVerification::GenesisPrevNotZero {
                    seq: e.seq,
                    detail: format!("expected prev_hash {}, got {}", prev_hash, e.prev_hash),
                }
            } else {
                ChainVerification::PrevHashMismatch {
                    seq: e.seq,
                    detail: format!("expected prev_hash {}, got {}", prev_hash, e.prev_hash),
                }
            };
        }
        let want = compute_entry_hash(e);
        if want != e.entry_hash {
            return ChainVerification::EntryHashMismatch {
                seq: e.seq,
                detail: format!("recomputed {}, stored {}", want, e.entry_hash),
            };
        }
        prev_hash = e.entry_hash.clone();
        prev_seq = e.seq as i64;
    }
    ChainVerification::Ok {
        count: entries.len(),
        last_entry_hash: prev_hash,
    }
}
